File: day03/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func treeCount(trees []string, dRow int, dCol int) (count int) {

	for row, col := 0, 0; row < len(trees); row, col = row+dRow, (col+dCol)%len(trees[0]) {
		if trees[row][col] == '#' {
			count++
		}
	}

	return
}

func partOne(trees []string) int {
	return treeCount(trees, 1, 3)
}

func partTwo(trees []string) int64 {
	product := int64(treeCount(trees, 2, 1))
	for i := 1; i < 9; i += 2 {
		product *= int64(treeCount(trees, 1, i))
	}
	return product
}

func main() {
	reader, _ := os.Open("input.txt")
	scanner := bufio.NewScanner(reader)
	trees := make([]string, 0)
	for scanner.Scan() {
		trees = append(trees, scanner.Text())
	}

	fmt.Println(partOne(trees))
	fmt.Println(partTwo(trees))
}

File: day03/solution.cpp
#include <iostream>
#include <fstream>
#include <set>
#include <vector>

using namespace std;

int slope_trees(const vector<vector<bool>> &trees, int d_row, int d_col) {
    int count = 0;
    for (int row = 0, col = 0; row < trees.size(); row += d_row, col = (col + d_col) % trees[0].size()) {
        count += trees[row][col];
    } 
    return count;
}

int part1(const vector<vector<bool>> &trees) {
    return slope_trees(trees, 1, 3);
}

int64_t part2(const vector<vector<bool>> &trees) {
    vector<pair<int, int>> slopes;
    int64_t product = slope_trees(trees, 2, 1);
    for (int i = 1; i < 9; i += 2) {
        product *= slope_trees(trees, 1, i);

    }
    return product;
}

int main() {
    ifstream in("input.txt");
    vector<vector<bool>> trees;
    string line;
    while (getline(in, line)) {
        if (line.empty()) {
            break;
        }
        vector<bool> row;
        for (char item : line) {
            row.push_back(item == '#');
        }
        trees.push_back(row);
    }
    cout << part1(trees) << endl << part2(trees) << endl;
}